File: incre-tests/synduce/list/alist_sum.f
Inductive List = nil Unit | cons {{Int, Int}, List};
Inductive CList = cnil Unit | single {Int, Int} | concat {CList, CList};

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    nil _ -> b
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
    \f: CList -> Compress List. \xs: CList.
    match xs with
      cnil _ -> nil unit
    | single x -> cons {x, nil unit}
    | concat {a, b} -> cat (f a) (f b)
    end
);

@Input xinp: Int;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> 
    let key = h.1 in
    let value = h.2 in
    if (== xinp key) then + value (f t) else (f t)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/alist_sum.ml
type list =
  | Nil of unit
  | Cons of (int * int) * list

type clist =
  | CNil of unit
  | Single of int * int
  | Concat of clist * clist

let rec cat a b =
  match a with
  | Nil _ -> b
  | Cons (h, t) -> Cons (h, cat t b)

@Input val xinp: int

val repr: clist -> list compress
let rec repr xs =
  match xs with
  | CNil _ -> Nil ()
  | Single x -> Cons (x, Nil ())
  | Concat (a, b) -> cat (repr a) (repr b)

let fi = function (x, _) -> x 
let se = function (_, x) -> x

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) ->
    if xinp == fi h then se h + spec t else spec t

let prog x = spec (repr x)